rfd/0151-database-permission-management.md

```diff
@@ -0,0 +1,221 @@
+- **Conciseness:** It should be relatively straightforward to write and create by hand, if necessary.
+- **Future-Proof:** The scheme should accommodate unforeseen changes in the future, such as the addition of additional levels within the hierarchy (e.g., schemas encompassing tables or columns within tables).
+- **Unambiguity:** If a wildcard character like `*` is allowed, it should have a clear and consistent meaning. Similarly, when specifying an object like "customers," it should be evident whether it refers to a table or a schema.
+
+### Wildcard Support
+
+Specifying permissions individually for each database object can be tedious and challenging to maintain. Wildcards provide a succinct and stable alternative, especially in the face of evolving schemas. It is encouraged that database engines support wildcard functionality for targets. This implementation may involve enumerating database objects and applying permissions to those matching specific patterns. Additionally, if a new object is created during a user session, the corresponding wildcard-based permission will not be automatically applied until a permission synchronization occurs. Implementing an automated permission synchronization triggered by schema changes can enhance the user experience.
+
+Unlike targets, permissions typically originate from a fixed set. While the inclusion of wildcards, as seen in the `ALL` permission, can still be beneficial, doing it at Teleport-level may overlap with database-level functionality, potentially diminishing utility.
+
+### Permission Lifecycle
+
+The feature is based on the automated database user creation feature, simplifying the analysis of permission lifetime.
+
+We define "permission synchronization" as the process of applying a predetermined set of permissions to the user. These permissions are calculated based on their current roles in conjunction with the existing database schema.
+
+1. Initially, when the user does not exist, they possess no permissions except for default ones.
+2. Upon connection, we create a user in a manner specific to the database. This involves applying a fixed set of roles to the user, as specified in `spec.allow.db_roles`.
+3. Permission synchronization occurs after the user is created or reactivated.
+4. Additional permission synchronizations can be triggered as necessary, such as on new connections, detected schema changes (which may be detected using database-specific methods), or based on a scheduled timer.
+5. For consistency, we ensure that new connections receive an equivalent set of permissions. Notably, permissions for other protocols or unsupported versions do not affect what is applied to the database. Changes in the database schema can lead to previously "identical" permission sets diverging, for example, when a new table is added. The permissions are equivalent if they are stable under schema changes.
+
+6. Once the last connection for the user is terminated, they are either removed or deactivated.
```

Comment:
A connection seems like a natural way to identify a database session, but in reality queries, subroutines, stored procedures, and likely other actions can continue to run after the user disconnects. We also need to consider things like MySQL Events which can run on a schedule: https://dev.mysql.com/doc/refman/8.0/en/events-overview.html

Have we considered how these potentially persistent conditions will be handled?

Reply:
Good question. I think this is relevant for the automated user provisioning feature too. In any case, once the last connection for a given user is dropped, all roles and permissions should be removed from the user. If possible, the user should be removed too. 

This means that any lingering processes running as a user should no longer have the ability to meaningfully interact with the database - as long as we are thorough with the removal of permissions.

Reply:
I don't believe this is true.  I believe in Postgres and MySQL the permissions are checked before execution starts.  And that deleting the user mid-execution will not impact already started operations.

Reply:
If we remove the permissions, but the database in question does not propagate those, then I'd say this is a problem (or a deliberate design) of the given database engine. I'm not sure what can we do more in this case, but I'm open for suggestions.

Reply:
The biggest thing we need to consider is possible interactions when adding the user back.  I am not aware of any reasons this will be a serious issue or of any security risks in this regard, but we should keep it in mind.

Besides that we need to document this concern.  This and https://github.com/gravitational/teleport/pull/33734#discussion_r1370674125 can likely be documented together: "You need to understand your database implementation, and how permissions are controlled and revoked" (in many more words)